storage/api/src/main/java/feast/storage/api/retrieval/HistoricalRetrievalResult.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright 2018-2020 The Feast Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.storage.api.retrieval;
+
+import com.google.auto.value.AutoValue;
+import feast.serving.ServingAPIProto.DataFormat;
+import feast.serving.ServingAPIProto.JobStatus;
+import java.io.Serializable;
+import java.util.List;
+import javax.annotation.Nullable;
+
+/** Result of a historical feature retrieval request. */
+@AutoValue
+public abstract class HistoricalRetrievalResult implements Serializable {
+
+  public abstract String getId();
+
+  public abstract JobStatus getStatus();
+
+  @Nullable
+  public abstract String getError();
+
+  @Nullable
+  public abstract List<String> getFileUris();
```

Comment:
Not a full review, but before forgetting I just wanted to carry over a point I had on #569 that's more appropriate to discuss in the wider context here:

> I feel a little reserved though about committing to the pattern of out-of-band file-based result data as API, i.e. `getFileUris()`. Not sure if large data sets will ever be streamed through RPC responses, but either of these options compromise the possibility for user jobs doing retrieval and processing the results to leverage data locality in systems where that could be achieved—see discussion starting at https://github.com/gojek/feast/issues/482#issuecomment-590827523

Reply:
@ches As you have rightly mentioned in #482, maintaining data locality requires a discussion on its own. I fear that our specific requirements at Gojek have steered us towards an approach that may be less than ideal in an HDFS + Spark world. 

You've mentioned that you have been working on a retrieval layer internally that bypasses our historical retrieval and presumably continues to maintain locality. 

I'd be happy to start a new issue on the requirements we (Gojek) currently have on the historical retrieval side, but ideally we'd have yours as well. Between the two we would cover a large swathe of use cases I believe.

Reply:
Sure, that's very reasonable, it's a complex subject but one that I think we've framed sufficiently to start that. I'm far from having a design proposal without requirements all laid out, but I can contribute to starting an RFC with problem statements.

For the sake of letting this PR move forward, maybe we can document `getFileUris` as a provisional API with a `@see` pointing to the new issue. In the worst case an implementation for which it doesn't make sense can throw an `UnsupportedOperationException` or the like and we can learn from use as this proceeds.

In our current approach with Hive, most likely we would not implement `HistoricalRetrievalResult` at all—we use Feast Ingestion but not Serving for historical, there is no `GetBatchFeatures` RPC call. It might be interesting to consider whether a similar call could be made mostly for Feast authorization purposes, metrics, etc., without expecting Serving to orchestrate the data staging. But, that's a matter for our new thread.

Reply:
>  It might be interesting to consider whether a similar call could be made mostly for Feast authorization purposes, metrics, etc., without expecting Serving to orchestrate the data staging. But, that's a matter for our new thread.

I'd really like to see if this is possible. 

By the way, how are you currently doing point in time correct joins? What triggers that execution?